federation/pkg/federation-controller/ingress/ingress_controller.go

```diff
@@ -240,6 +346,131 @@ func (ic *IngressController) reconcileIngressesOnClusterChange() {
+
+	clusterIngressUID, clusterIngressUIDExists := cluster.ObjectMeta.Annotations[uidAnnotationKey]
+	configMapUID, _ := configMap.Data[uidKey]
+	if !clusterIngressUIDExists || clusterIngressUID == "" {
+		ic.updateClusterIngressUIDToMasters(cluster, configMapUID) // Second argument is the fallback, in case this is the only cluster, in which case it becomes the master
+		return
+	}
+	if configMapUID != clusterIngressUID { // An update is required
+		configMap.ObjectMeta.Annotations[uidAnnotationKey] = clusterIngressUID
+		operations := []util.FederatedOperation{{
+			Type:        util.OperationTypeUpdate,
+			Obj:         &configMap,
+			ClusterName: cluster.Name,
+		}}
+		glog.V(4).Infof("Calling federatedConfigMapUpdater.Update() - operations: %v", operations)
+		err := ic.federatedConfigMapUpdater.Update(operations, ic.updateTimeout)
+		if err != nil {
+			configMapName := types.NamespacedName{Name: configMap.Name, Namespace: configMap.Namespace}
+			glog.Errorf("Failed to execute update of ConfigMap %q on cluster %q: %v", configMapName, cluster.Name, err)
+			// TODO: ic.deliverConfigMap(configMap, ic.configmapReviewDelay, true)
+		}
+	}
+}
+
+/*
+  getMasterCluster returns the cluster which is the elected master w.r.t. ingress UID, and it's ingress UID.
+  if there is no elected master cluster, and error is returned.
+  All other clusters must use the ingress UID of the elected master.
+*/
+func (ic *IngressController) getMasterCluster() (master *federation_api.Cluster, ingressUID string, err error) {
+	clusters, err := ic.ingressFederatedInformer.GetReadyClusters() // TODO, get all clusters, not just ready ones.
+	if err != nil {
+		glog.Errorf("Failed to get cluster list: %v", err)
+		return nil, "", err
+	}
+
+	for _, c := range clusters {
+		UID, exists := c.ObjectMeta.Annotations[uidAnnotationKey]
+		if exists && UID != "" { // Found the master cluster
+			return c, UID, nil
+		}
+	}
+	return nil, "", fmt.Errorf("Failed to find master cluster with annotation %q", uidAnnotationKey)
+}
+
+/*
+  updateClusterIngressUIDToMasters takes the ingress UID annotation on the master cluster and applies it to cluster.
+  If there is no master cluster, then fallbackUID is used (and hence this cluster becomes the master).
+*/
+func (ic *IngressController) updateClusterIngressUIDToMasters(cluster *federation_api.Cluster, fallbackUID string) {
+	masterCluster, masterUID, err := ic.getMasterCluster()
+	if err != nil {
+		if masterCluster.Name != cluster.Name { // If we are the master, no update needed
+			cluster.ObjectMeta.Annotations[uidAnnotationKey] = masterUID
+			if _, err = ic.federatedApiClient.Federation().Clusters().Update(cluster); err != nil {
+				glog.Errorf("Failed to add master ingress UID annotation (%q = %q) from master cluster %q to cluster %q, will try again later: %v", uidAnnotationKey, masterUID, masterCluster.Name, cluster.Name, err)
```

Comment:
No, once a master cluster has been chosen once, it's UID is applied to all other clusters (in the federation etcd, and thence to the cluster's ingress controller configmaps), and if any one of them becomes the master in future, the UID will remain constant.  This only deals with the case where none of the clusters (in the federation etcd) know what the UID is (e.g. when only one cluster is registered), in which case the one from that (now master) cluster is used.

It would be better to persist this in one central place rather than in all of the registered clusters, but place does not exist yet.  I will add an issue to track improving this.